smolyay/samples.py

```diff
@@ -0,0 +1,207 @@
+import abc
+import sys
+import numpy
+
+
+class UnidimensionalPointSet(abc.ABC):
+    """Set of unidimensional points
+
+    A set of unique unidimensional points to be used as the sampling points
+    in a tensor product grid or sparse grid. These unidimensional points are
+    associated with certain polynomial families, and the combination of
+    unique points and polynomial families are the basis for different types
+    of quadrature rules that are used in numerical integration and in
+    approximation.
+    The natural domain of the points specified by :param:``natural_domain`` is
+    the domain in which all the possible points in the set lies within.
+    ``number_points`` is the number of points that the UnidimensionalPointSet
+    is expected to store.
+    ``domain`` is the domain of the points generated by the point set.
+    """
+
+    def __init__(self):
+        self._points = []
+
+    @property
+    @abc.abstractmethod
+    def domain(self):
+        """list: Domain the sample points come from."""
+        pass
+
+    @property
+    def points(self):
+        """list: Points stored by the set."""
+        return self._points
+
+    @abc.abstractmethod
+    def generate_points(self, num_points):
+        r"""Generating the points
+
+        An abstract method for generating the points stored by the point set.
+        Called when the number of points requested is larger than the number
+        of points available.
+
+        Parameters
+        ----------
+        num_points : int
+            The number of points to generate
+
+        Returns
+        -------
+        self
+            The UnidimensionalPointSet
+        """
+        pass
+
+
+class ClenshawCurtisPointSet(UnidimensionalPointSet):
+    r"""Set of unidimensional points for Clenshaw Curtis sampling
+
+    The :attr:`points` for this interpolation scheme are the extrema of the
+    Chebyshev polynomials of the first kind on the domain :math:`[-1, 1]`:
+
+    .. math::
+
+        x_i^* = -\cos(\pi i/n), i = 0,...,n
+
+    For the special case :math:`n = 0`, there is only one point :math:`x_0^* = 0`.
+    """
+
+    def __init__(self):
```

Comment:
This constructor needs an argument, like `degree` or `N` representing the degree of the polynomial these points come from.
```suggestion
    def __init__(self, degree):
```